src/generic/stage2/numberparsing.h

```diff
@@ -110,61 +102,57 @@ simdjson_really_inline bool compute_float_64(int64_t power, uint64_t i, bool neg
   // is equal to
   //  floor(log(5**power)/log(2)) + power
   //
-  // The 65536 is (1<<16) and corresponds to 
+  // The 65536 is (1<<16) and corresponds to
   // (65536 * power) >> 16 ---> power
   //
-  // ((152170 * power ) >> 16) is equal to 
-  // floor(log(5**power)/log(2)) 
+  // ((152170 * power ) >> 16) is equal to
+  // floor(log(5**power)/log(2))
   //
-  // Note that this is not magic: 152170/(1<<16) is 
+  // Note that this is not magic: 152170/(1<<16) is
   // approximatively equal to log(5)/log(2).
-  // The 1<<16 value is a power of two; we could use a 
+  // The 1<<16 value is a power of two; we could use a
   // larger power of 2 if we wanted to.
   //
   int64_t exponent = (((152170 + 65536) * power) >> 16) + 1024 + 63;
-  
+
 
   // We want the most significant bit of i to be 1. Shift if needed.
   int lz = leading_zeroes(i);
   i <<= lz;
+
+
+  // We are going to need to do some 64-bit arithmetic to get a  precise product.
+  // We use a table lookup approach.
+  // It is safe because
+  // power >= smallest_power
+  // and power <= largest_power
+  // We recover the mantissa of the power, it has a leading 1. It is always
+  // rounded down.
+  //
   // We want the most significant 64 bits of the product. We know
   // this will be non-zero because the most significant bit of i is
   // 1.
-  value128 product = full_multiplication(i, factor_mantissa);
-  uint64_t lower = product.low;
-  uint64_t upper = product.high;
-
-  // We know that upper has at most one leading zero because
-  // both i and  factor_mantissa have a leading one. This means
-  // that the result is at least as large as ((1<<63)*(1<<63))/(1<<64).
-
-  // As long as the first 9 bits of "upper" are not "1", then we
-  // know that we have an exact computed value for the leading
-  // 55 bits because any imprecision would play out as a +1, in
-  // the worst case.
-  if (simdjson_unlikely((upper & 0x1FF) == 0x1FF) && (lower + i < lower)) {
-    uint64_t factor_mantissa_low =
-        mantissa_128[power - FASTFLOAT_SMALLEST_POWER];
-    // next, we compute the 64-bit x 128-bit multiplication, getting a 192-bit
-    // result (three 64-bit values)
-    product = full_multiplication(i, factor_mantissa_low);
-    uint64_t product_low = product.low;
-    uint64_t product_middle2 = product.high;
-    uint64_t product_middle1 = lower;
-    uint64_t product_high = upper;
-    uint64_t product_middle = product_middle1 + product_middle2;
-    if (product_middle < product_middle1) {
-      product_high++; // overflow carry
-    }
-    // We want to check whether mantissa *i + i would affect our result.
-    // This does happen, e.g. with 7.3177701707893310e+15.
-    if (((product_middle + 1 == 0) && ((product_high & 0x1FF) == 0x1FF) &&
-         (product_low + i < product_low))) { // let us be prudent and bail out.
+  const uint32_t index = 2 * uint32_t(power - smallest_power);
```

Comment:
I'm sure the compiler will (probably) optimize it, but it's worth noting here that Intel addressing modes can handle something like (small power of 2) * X + Y + Constant all in a single address lookup. It's not necessary to hyperoptimize just yet, but since the constant is applied *before* the multiplication instead of after, and this calculation is indirect, it's worth a note to check the assembly for `power_of_five_128[index]` and `power_of_five_128[index+1]` to see if they are single instructions.

Reply:
@jkeiser I do not remember the details, but there are definitively cases on some processors where loads from simple addressing schemes are faster. I *think* that Intel processors can issue two loads per cycle and one store, if and only if the addressing is simple. 

Even if that is wrong, Travis reminds us that there is definitively a latency cost on intel processors...
https://stackoverflow.com/a/51664696


I'll add a remark, however, so that we can revisit this later.